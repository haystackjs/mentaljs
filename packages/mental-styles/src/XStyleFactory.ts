export interface XStyleFactory {
    createStyle: (style: any) => string;
}

class XStyleFactoryHolder {
    private _factory?: XStyleFactory;

    get factory(): XStyleFactory {
        if (!this._factory) {
            throw Error('XStyleFactory is not set');
        }
        return this._factory;
    }

    registerFactory(factory: XStyleFactory) {
        // Should we throw an error or we just don't care?
        this._factory = factory;
    }
}

export const XStyleFactoryRegistry = new XStyleFactoryHolder();