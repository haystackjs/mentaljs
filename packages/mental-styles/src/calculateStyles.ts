import { XStyleFactoryRegistry } from './XStyleFactory';
import { XStyles } from './XStyles';
import { prepareStyles } from './utils/prepareStyles';
import { extractModifier } from './utils/extractModifier';
import { stylesMap } from './utils/stylesMap';

// Our styles cache. Can it be faster?
const stylesCache = new Map<string, string>();

export function calculateStyles(styles: XStyles, selected: boolean = false) {
    const factory = XStyleFactoryRegistry.factory;

    // Load styles
    let src = styles;
    let srcHover = extractModifier('hover', styles);
    if (selected) {
        src = prepareStyles({ ...src, ...extractModifier('selected', styles) });
        srcHover = prepareStyles({ ...srcHover, ...extractModifier('selectedHover', styles) });
    }
    src = prepareStyles(src);
    srcHover = prepareStyles(srcHover);

    // Building CSS class names
    let css: string[] = ['x'];
    for (let k of Object.keys(src)) {
        let v = src[k];
        let key = k + ': ' + v;
        if (!stylesCache.has(key)) {
            stylesCache.set(key, factory.createStyle({ k: v }));
        }
        css.push(stylesCache.get(key)!);
    }
    for (let k of Object.keys(srcHover)) {
        let v = srcHover[k];
        let key = 'hover-' + k + ': ' + v;
        if (!stylesCache.has(key)) {
            stylesCache.set(key, factory.createStyle({
                '&:hover, &:focus': { k: v }
            }));
        }
        css.push(stylesCache.get(key)!);
    }

    return css.join(' ');
}
