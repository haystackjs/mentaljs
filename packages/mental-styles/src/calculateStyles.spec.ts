import { XStyleFactoryRegistry } from './XStyleFactory';
import { css } from 'glamor';
import { calculateStyles } from "./calculateStyles";
XStyleFactoryRegistry.registerFactory({
    createStyle: (styles) => {
        return css(styles).toString()
    }
});

describe('calculateStyles', () => {
    it('should caluclate style', () => {
        let st = calculateStyles({ extractLayer: true });
        expect(st).toMatchSnapshot();
    })
});