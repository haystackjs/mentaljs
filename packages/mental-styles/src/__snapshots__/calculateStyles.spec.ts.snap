// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateStyles should caluclate style 1`] = `"x css-17eot9t"`;
